mchf-eclipse/drivers/ui/ui_driver.c

```diff
@@ -6940,6 +6960,82 @@ void UiDriver_TaskHandler_HighPrioTasks()
 
 }
 
+static void UiDriver_HandleUSB_Keyboard()
+{
+#ifdef USE_USBKEYBOARD
+    HID_KEYBD_Info_TypeDef *k_pinfo;
+    static uint8_t keys_buffer[sizeof(k_pinfo->keys)] = { 0 };
+
+    if(USBH_HID_GetDeviceType(&hUsbHostHS) == HID_KEYBOARD)
+    {
+        k_pinfo = USBH_HID_GetKeybdInfo(&hUsbHostHS);
+
+        if(k_pinfo != NULL)
+        {
+            for( uint32_t idx = 0; idx < sizeof(k_pinfo->keys); idx++ )
+            {
+                bool is_exist_in_buffer = false;
+                for( uint32_t i = 0; i < sizeof(k_pinfo->keys); i++ )
+                {
+                    if ( keys_buffer[i] == k_pinfo->keys[idx])
+                    {
+                        is_exist_in_buffer = true;
+                    }
+                }
+                if ( is_exist_in_buffer == true )
+                {
+                    continue;
+                }
+                else
```

Comment:
> At least for me there has never been an issue with "bouncing" keys.

it's not "bouncing"
As I asked, try to press two buttons at the same time, what do you see?
Or hold one and press another while you holding first one. 

> The data is read from a fifo filled by the USB host controller code. 

Maybe keyboards different but according to HID specification it's not FIFO
![](https://s8.hostingkartinok.com/uploads/images/2019/01/09152a5a67019e94662b8192b362bed6.jpg)
order is not determinated and this is responsobility of the host to figure out... 

It could be done more elegant if order was deterministic, but... save to fallow HID specification I think.

Sure, I will comment later...

Reply:
I see. What your code does is the following: If I press a `M` and keep it pressed, and then press an `A` your code would generate a M and then an A as output. But I thought this is wrong. However, I tried it on my PC and yes, it does generate n M and then an A, so your code is not wrong but I was. Never noticed this. 

However, I still think we should point the reader of this code into the right direction with a comment. Otherwise it is difficult to understand what is the purpose of this piece of code.